### PR TITLE
Basic Project Setup + Migrating Code from PySDK PR #298

### PR DESCRIPTION
**Pull request by tzaffi**
# TEAL Blackbox in Graviton

This PR takes over from the now defunct [PySDK PR #298](https://github.com/algorand/py-algorand-sdk/pull/298) 

## A Place to Test Our Python Repo Best Practices

Since this is a brand new repo this is a good place to set some new conventions that our other Python repos should eventually follow also. This PR does not claim to have achieved such a state (especially in light of issues #4 and #7).

Some conventions are implicitly advanced in this PR:
* no `requirements.txt`:
  * instead, add an `extras_require` (e.g. `extras_require={"test": "pytest==7.1.1"},`) in `setup.py`
  * install the test-dependencies with `pip install -e.[test]`
* pin internal algorand dependencies using version tags
* (For projects that use github actions) Use `act` to simulate github actions locally
* Use `Makefile` to summarize basic setup/testing commands, and incorporate these commands in CI configs
* ~`tests` should contain all unit and integration tests (as opposed to pairing `*_test.py` files next to the file they are testing)~ No longer considered a "best practice"
* test files should not be installed by default (previous bullet point makes this goal easier)
* linter configurations should go into `setup.cfg` whenever possible
* lint with `black --check` and `flake8`

### Testing

```sh
make integration-test
```

### Simulating Github Actions Locally (on a Mac)
```sh
make mac-gh-simulate
````

### CI / Best Practices
- [ ] #2 
- [ ] #4 
- [ ] #6 
- [ ] #7 

### Features / Improvements
- [ ] #3 
- [ ] #5 

### Migrate Sandbox Code from PySDK and Run All Tests

Testing is cribbing off of PyTeal's ["Run blackbox tests in CI" PR](https://github.com/algorand/pyteal/pull/260)

* [First successful docker build](https://github.com/algorand/graviton/runs/5685924131?check_suite_focus=true)
  * TIME: 35 minutes
  * Compare with 35 minutes for PyTeal's [first uncached build](https://github.com/algorand/pyteal/runs/5676255363?check_suite_focus=true)

* [First cached docker build](https://github.com/algorand/graviton/runs/5686336452?check_suite_focus=true)
  * TIME: 4 minutes
  * Compare with 5 minutes for [a cached build](https://github.com/algorand/pyteal/runs/5676796537?check_suite_focus=true) on PyTeal

### CODE REVIEW

**Comment by tzaffi** [comment]
# Continuation of the most important thread from the defunct PySDK PR

@jasonpaulos  made 4 important comments, and this is how I previously responded. Below, I am updating my responses.
image

File: blackbox/dryrun.py
Comment (tzaffi) [inline blackbox/dryrun.py:1]: With this relatively old (and slightly refactored code) coming from PySDK, we should consider removing -or at least deprecating- [PySDK's testing folder](https://github.com/algorand/py-algorand-sdk/tree/develop/algosdk/testing). This python functionality is definitely not reproduced in our other SDK's.

@jasonpaulos @winder

File: tests/integration/dryrun_mixin_docs_test.py
Comment (tzaffi) [inline tests/integration/dryrun_mixin_docs_test.py:3]: New test, based on old devrel docs. Instroduced mainly to verify that refactoring of `DryRunTestMixin` did not break existing functionality.

File: Makefile
Comment (tzaffi) [inline Makefile:47]:
```suggestion
	# relax exit code condition because indexer returns 500 when last-round = 0
```

File: Makefile
Comment (tzaffi) [inline Makefile:40]:
```suggestion
# assumes act is installed, e.g. via `brew install act`:
```

File: setup.py
Comment (tzaffi) [inline setup.py:24]: I'm stuck with this dependency because of our C.I. process which uses ubuntu-20 and on which it seems rather difficult to get python 3.10 going on.

Comment (tzaffi) [reply setup.py:24]: @michaeldiamant ^^^ this might be an issue for our goal of getting PyTEAL to 3.10 as well

File: Makefile
Comment (michaeldiamant) [inline Makefile:25]: @tzaffi Asking having briefly reviewed the PR:  What is Mac-specific about these build commands?

Though I'm unfamiliar with some of the tooling (e.g. tox), at first glance, I'm _guessing_ it's label's _Mac only_ because that's the only environment tested locally.

My default assumption is that the build is _not_ OS-specific.

File: blackbox/dryrun.py
Comment (tzaffi) [reply blackbox/dryrun.py:1]: @barnjamin is introducing `algosdk/dryrun_results.py` which is a separate module outsidef of  `algosdk/testing/`. It's reasonable that the code living in `algosdk/testing/` would be refactored to use the code now being introduced, but the bigger question is it should even exist in PySDK?

File: Makefile
Comment (tzaffi) [reply Makefile:25]: My original intent with the `mac-*` prefix was to differentiate between commands that I know will be successful on github, and ones that aren't. Also, by using `mac` I'm hinting at the fact that I only know really that these can work on our Algorand laptops and make no guarantee for other os'es. But the naming is a bit jarring. Do you think `local-*` would be better?

Comment (michaeldiamant) [reply Makefile:25]: @tzaffi Got it - Unless we know about OS-specific limitations, I'd opt for something like `local`.

I totally see your concern that we're not exercising the commands on multiple OSes, so we can't be _certain_.  But, I'm hoping we can cross that bridge should it arise rather than starting from the OS-specific assumption.

Ultimately, the discussion points us back towards unifying the local and CI builds.  If you feel strongly, I am OK to keep as is.  There's bigger fish to fry.

File: setup.py
Comment (michaeldiamant) [reply setup.py:24]: @tzaffi I opened a child PR to help move to 3.10:  https://github.com/algorand/graviton/pull/8.  Can you have a look and let me know if you're up for merging it?

Note:  The build did _not_ find a cached sandbox.  So, I'm leaving the message ahead of the build completing.  Local testing `act` suggests it works.  I'll double check tomorrow.  🙏 

Aside:  Is it possible to configure an upper bound for supported Python versions?
* It's _not_ a pressing concern.  
* My passing thought is:  If the build claims to support newer versions, then the CI build ought to enforce it (e.g. matrix build).  For example, consider Python 3.11 coming out, but not being built by CI.

File: tests/integration/dryrun_mixin_docs_test.py
Comment (michaeldiamant) [inline tests/integration/dryrun_mixin_docs_test.py:396]: @tzaffi Minor:  I didn't inspect closely though it seems `pprint` results in _a lot_ of stdout during test runs.  Leaving the note to sanity check if it's straight-forward to reduce output (e.g. via log levels) at least when tests pass?

File: setup.py
Comment (tzaffi) [reply setup.py]: This is nice. I'll work to incorporate this into the PR!

File: Makefile
Comment (michaeldiamant) [reply Makefile:21]: @tzaffi Thanks for the naming change.

File: setup.py
Comment (tzaffi) [reply setup.py]: FYI: 
* I merged #8  in. Works great!
* I tried naively adding "3.11" to the github-actions python matrix, and it didn't work. It appears [not to yet be supported](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#python).

File: Makefile
Comment (tzaffi) [inline Makefile:6]: @jasonpaulos - per our conversation, I'm recommending going with the approach of having a special "test" requirements using `setup.py`'s `require_extras` dict.

Some brief research shows that there is [PEP 621](https://peps.python.org/pep-0621/#example) will give a different solution in the future, but AFAIK no version of `pip` yet supports this. Therefore, I'm going with the next best option, in my opinion, which is [discussed here](https://discuss.python.org/t/is-it-preferable-to-add-test-dependensies-as-an-extra-require-instead-of-tox-ini/4984?filter=summary).

File: Makefile
Comment (tzaffi) [inline Makefile:24]:
```suggestion
###### Local Only ######
```

File: Makefile
Comment (tzaffi) [inline Makefile:53]:
```suggestion
.PHONY: pip-publish pip-test unit-test gh-blackbox

```

File: setup.py
Comment (tzaffi) [inline setup.py:14]: Going aggressive here with Python 3.10. Open to other suggestions.

File: tests/integration/doc_examples_test.py
Comment (tzaffi) [inline tests/integration/doc_examples_test.py:1]: These tests are validating that the README's examples are correct.

File: tests/integration/dryrun_mixin_docs_test.py
Comment (tzaffi) [reply tests/integration/dryrun_mixin_docs_test.py:3]: @algorandskiy - Here's what I was talking about today.

File: Makefile
Comment (michaeldiamant) [inline Makefile:53]: @tzaffi `pip-publish` appears among `.PHONY` targets, but nowhere else.  I'm wondering if it's stale?

If not stale, can you help me understand the target's intent?

File: graviton/invariant.py
Comment (michaeldiamant) [inline graviton/invariant.py:44]: @tzaffi Minor:  IDE hints that `property` shadows a built-in name.  Unless you feel strongly, it seems prudent to rename to avoid unintended side-effects.  

If you prefer to keep as is, then let's resolve the thread.

File: graviton/dryrun.py
Comment (michaeldiamant) [inline graviton/dryrun.py:5]: @tzaffi Minor:  I _think_ the import is _not_ needed.  I point it out because a test dependency felt surprising here.

```suggestion

```

File: graviton/dryrun.py
Comment (michaeldiamant) [inline graviton/dryrun.py:211]: @tzaffi Sanity check:  Based on indents, it seems `add_found` is _not_ lexically available.  Can you double check me?

As I see it, the highlighted statement is _outside_ the for-loop.  If that's accurate, then I'm pointing out the concern because the check won't work as intended.

File: graviton/dryrun.py
Comment (michaeldiamant) [inline graviton/dryrun.py:116]: @tzaffi Minor - Should be safe to remove the definition.  It's redefined/reassigned in the same lexical scope.

```suggestion

```

File: tests/clients.py
Comment (michaeldiamant) [inline tests/clients.py:16]: @tzaffi Minor:  Friendly pointer to commented out code.  Can it be removed?

File: Makefile
Comment (michaeldiamant) [inline Makefile:23]: @tzaffi For my understanding:  How is graviton published for downstream consumption (e.g. pyteal)?

I suspect the mechanism exists here, but I'm unfamiliar with the tooling.  I raise the question because none of the `make` targets seem to correlate.

File: setup.py
Comment (michaeldiamant) [reply setup.py:14]: @tzaffi I'd like to go to Python 3.10 too.  Presumably, requiring 3.10 here implies the downstream PyTeal PR (e.g. https://github.com/algorand/pyteal/pull/249) must also use 3.10 as a minimum version?

If that's the case, then some work will be needed to bring PyTeal to 3.10 minimum version.  I think the change makes sense.  I'm asking the question to gauge dependencies.

File: README.md
Comment (michaeldiamant) [inline README.md:4]: @tzaffi @jasonpaulos What support guarantees does graviton provide?

* My default preference is to add a _brief_ README disclaimer like:  Graviton is subject to change and makes no backwards compatibly guarantees.
* I think a broader conversation exists for support guarantees beyond graviton.  I'm trying to avoid the broader conversation while ensuring we retain flexibly in graviton.  I'm flexible on discussion specific to graviton.

File: Makefile
Comment (tzaffi) [reply Makefile:53]: Good catch. `.PHONY` ensures that make targets which don't produce a file always run, even in the case that the file with the same name as the target exists. At some point during bootstraping this repo, my make commands became impotent and introducing the `.PHONY`'s fixed it. I'm not sure which file was being shadowed. However, we should _never_ be having shadowing problems on github, because we should never check a shadowing file into git. So I'll remove for now. If the problem ever pops again, I'll investigate it a little more deeply.

File: graviton/invariant.py
Comment (tzaffi) [reply graviton/invariant.py]: Another good catch!

File: graviton/dryrun.py
Comment (tzaffi) [reply graviton/dryrun.py:5]: gonna add `flake8` to nip this in the bud

File: setup.cfg
Comment (tzaffi) [inline setup.cfg:2]:
```suggestion
ignore = E501, W503

```

File: setup.cfg
Comment (tzaffi) [inline setup.cfg:1]: ignore line lengths (let `black` take care of that) and another conflict with `black` over binary operator line break placement

File: setup.py
Comment (tzaffi) [reply setup.py:14]: This doesn't require anything greater than 3.8. Maybe I'm safe lowering the version here, since all the dependencies getting installed have specific pinned versions. LMK

File: graviton/dryrun.py
Comment (tzaffi) [reply graviton/dryrun.py:210]: Good catch.

Inheriting old bad code here from pysdk. I'll correct and let the integration tests be the judge.

File: graviton/dryrun.py
Comment (tzaffi) [reply graviton/dryrun.py:115]: 👍

File: tests/clients.py
Comment (tzaffi) [reply tests/clients.py:16]: ok... the reason I was keeping it around is too speculative at this point

File: Makefile
Comment (tzaffi) [reply Makefile:33]: It's confusing. But basically as long as:
* you use `pip` to install from this python project (whether from PyPi for Github)
* you have the correct project structure (in particular the `./graviton/` directory has to have all the `__init__.py` files
* you have the correct `setup.py` (using `find_modules()`) 
then everything works automagically.

But that's exactly why I'm not particularly enamored. 

I can see the point of **convention over configuration** if the convention wasn't so clunky...

File: README.md
Comment (tzaffi) [reply README.md:4]: 👍

File: setup.py
Comment (michaeldiamant) [reply setup.py]: @tzaffi  My primary focus is on bringing the PR to a merge, so I don't want to hold up things.

So, I'm on-board with the change.  But, the following question arises:  Is there a reason _not_ to test the integration tests with the same version matrix?

If there is a rationale, I'm thinking an inline comment to the build is appropriate to benefit future maintainers.

Comment (tzaffi) [reply setup.py]: After our IRL discussion and further thought, I believe an integration python test matrix is appropriate. Will modify accordingly.

File: Makefile
Comment (michaeldiamant) [reply Makefile:33]: @tzaffi Thanks for the follow up here.  Resolving.

File: .github/workflows/build.yml
Comment (michaeldiamant) [inline .github/workflows/build.yml:32]: @tzaffi Minor:  Friendly pointer to commented out code.

```suggestion

```

File: .github/workflows/build.yml
Comment (michaeldiamant) [inline .github/workflows/build.yml:46]: @tzaffi Minor:  Friendly pointer to commented out code.

```suggestion
```

**Review by michaeldiamant — APPROVED** [review]
@tzaffi Thanks for tying together a number of threads to bring graviton to life.  ☕ 

For review completeness:  I'm on-board to try the proposed versioning scheme though I have lingering concerns.  Since we can more freely make changes here, let's try it out and assess over time.

Rationale:
* It's possible unicode characters interact poorly with other systems (e.g. build systems) and/or are difficult to parse when _not_ presented as emojis (e.g. `git tag --list` vs `git --no-pager tag --list`).
* The convention does _not_ support signaling when a breaking change is made.  The lack of signaling makes it harder to know when it's safe to upgrade.

File: .github/workflows/build.yml
Comment (tzaffi) [reply .github/workflows/build.yml:46]: 👍

File: Makefile
Comment (jasonpaulos) [inline Makefile:4]: I'd rather not have build tool versions constantly shifting under our feet

**Comment by tzaffi** [comment]
> @tzaffi Thanks for tying together a number of threads to bring graviton to life. ☕
> 
> For review completeness: I'm on-board to try the proposed versioning scheme though I have lingering concerns. Since we can more freely make changes here, let's try it out and assess over time.
> 
> Rationale:
> 
> * It's possible unicode characters interact poorly with other systems (e.g. build systems) and/or are difficult to parse when _not_ presented as emojis (e.g. `git tag --list` vs `git --no-pager tag --list`).
> * The convention does _not_ support signaling when a breaking change is made.  The lack of signaling makes it harder to know when it's safe to upgrade.

We should keep the new versioning scheme at least through April 1, 2022

File: Makefile
Comment (jasonpaulos) [reply Makefile:7]: I can accept this, though I'd suggestion `development` or similar instead of just `test`. The dev process involves more than just tests, there are also linters, formatters, etc. that help us write code.

File: .github/workflows/build.yml
Comment (tzaffi) [reply .github/workflows/build.yml]: 👍

File: setup.cfg
Comment (jasonpaulos) [reply setup.cfg:1]: Why is there a `setup.cfg` _and_ `setup.py`? Is there any other place to put this information?

**Review by jasonpaulos — COMMENTED** [review]
Comments that didn't make sense on a particular line:
* I think the animal emoji versioning scheme only make sense if it's mapped directly to a semver version (e.g. 🐈 = 0.1.0). I think it's good marketing, but it should not replace semver and traditional versioning.
* I'm in favor of having `*_test.py` unit tests that live in the same directory as implementation code. They are very convenient and act as a constant reminder to write unit tests for your code.

I only shared the above two comments because the context of this repo seems to be to establish best practices for future repos. If that were not the case, my opinions on these issue would not be very strong.

File: setup.cfg
Comment (tzaffi) [reply setup.cfg:1]: I ended up needing the `setup.cfg` for the [flake8 configuration](https://flake8.pycqa.org/en/latest/user/configuration.html#configuration-locations). I couldn't see where in `setup.py` I would put it.

Comment (tzaffi) [reply setup.cfg:1]: Seems like it's still going to be used per [PEP 518](https://peps.python.org/pep-0518/#sticking-with-setup-cfg) (even if there's not much enthusiasm in that PEP for it)

Comment (tzaffi) [reply setup.cfg:1]: And flake8 has an [issue outstanding](https://github.com/PyCQA/flake8/issues/234) to support `pyproject.toml`

File: Makefile
Comment (tzaffi) [reply Makefile:4]: 👍

File: setup.cfg
Comment (jasonpaulos) [reply setup.cfg:1]: It seems like `tox.ini` and `.flake8` are bother alternatives to putting this in `setup.cfg`. Either seems preferable to defining both a `setup.py` and `setup.cfg`, since I have concerns about having both of those files in the same package

Comment (tzaffi) [reply setup.cfg:1]: 👍

File: Makefile
Comment (tzaffi) [reply Makefile]: 👍

**Review by jasonpaulos — APPROVED** [review]

**Pull request merged**
